src/utils/solidity/prepareTransaction.ts

```diff
@@ -15,25 +33,60 @@ export interface PreparedTransaction {
 
 export const prepareTransaction = async (
   transaction,
-  options?,
-  environment = getGlobalEnvironment(),
+  optionsOrEnvironment: OptionsOrCallback | Environment,
+  maybeEnvironment = getGlobalEnvironment(),
 ): Promise<PreparedTransaction> => {
   const encoded = transaction.encodeABI();
 
-  if (options === undefined || options.gas === undefined) {
-    try {
-      transaction.gasEstimation = await transaction.estimateGas({
+  const environment = isEnvironment(optionsOrEnvironment)
+    ? optionsOrEnvironment
+    : maybeEnvironment;
+
+  const options = isEnvironment(optionsOrEnvironment)
+    ? {
+        amguPayable: false,
         from: environment.wallet.address.toString(),
-      });
-    } catch (e) {
-      throw new Error(
-        `Gas estimation (preflight) failed for ${
-          transaction.name
-        }(${transaction.arguments.join(', ')}): ${e.message}`,
-      );
-    }
-  } else {
-    transaction.gasEstimation = options.gas;
+      }
+    : {
+        amguPayable: false,
+        from: environment.wallet.address.toString(),
+        ...(typeof optionsOrEnvironment === 'function'
+          ? optionsOrEnvironment(environment)
+          : optionsOrEnvironment),
+      };
+
+  const maxGasPrice = multiply(
+    toBI(defaultOptions.gasLimit),
+    toBI(defaultOptions.gasPrice),
+  );
+
+  const balance = await environment.eth.getBalance(options.from);
+  const maxAmguInEth = subtract(toBI(balance), maxGasPrice);
+
+  // We don't know the amgu price at this stage yet, so we just send all
+  // available ETH for the gasEstimation. This should throw if amgu price
```

Comment:
can we find a way to know the amgu price before that ? Dont like the idea of sending the entire balance ; doesnt feel right

Reply:
@schmidsi

Reply:
ok gotcha nice